### ISSUE TITLE
hibernation not being automatic using a swap file

### ISSUE THREAD

**Issue by beh-10257** [NixOS/nixpkgs#276374]
**Describe the bug**
systemctl can't find the file to hibernate to
so yeah no hibernation

**Expected behavior**

hibernation working

```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.68, NixOS, 24.05 (Uakari), 24.05.20231219.54aac08`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"home-manager, nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Additional context**
https://discourse.nixos.org/t/btrfs-swap-not-enough-swap-space-for-hibernation/36805
my whole filesystem is encrypted excluding the boot partition

![image](https://github.com/NixOS/nix/assets/77834479/f2806506-eb80-41d4-b833-5323dc98c2b7)
this shows that the swap actually works
![image](https://github.com/NixOS/nix/assets/77834479/65ea4dc1-59a3-490d-bce1-a06a1213d506)
this shows specifically the errors

[my journalctl -b 0 --system](https://github.com/NixOS/nix/files/13759989/broken_hibernation.log)

basically copied from https://github.com/NixOS/nix/issues/9660
also I don't really know if this is a btrfs or a kernel or systemd issue so not mentioning anyone

**Comment by charmoniumQ** [comment]
Sounds like a systemd issue that was fixed in v255

https://github.com/systemd/systemd/issues/30599

**Comment by YHNdnzj** [comment]
> Sounds like a systemd issue that was fixed in v255

To clarify: The newly-added support was for _automatically_ detecting the swapfile offset on btrfs. Manually setting `resume=` and `resume_offset=` kernel parameters should work regardless of the version of systemd.

**Comment by charmoniumQ** [comment]
See also `systemd: 254.6 -> 255` #265951

**Comment by beh-10257** [comment]
I don't know if its really suppposed to be an automatic process but as far as I am concerned even with upgrading to systemd 255 from nixpkgs-master still same issues
![image](https://github.com/NixOS/nixpkgs/assets/77834479/5a91b340-42d0-4078-b282-e5579cf4552d)

basically is this supposed to be an automatic process like regesting where the file is and continuing from there
for context
it now doesn't crash when I systemctl hibernate
but when I open my system nothing is restored so its as if I did no hibernation
@charmoniumQ @YHNdnzj

**Comment by YHNdnzj** [comment]
IIRC NixOS uses a different build of systemd that doesn't come with EFI support in initrd. So automatic resuming may not work.

I'm not a NixOS user, so I'd suggest going to distro maintainers for help.

**Comment by beh-10257** [comment]
FWIW I'll tag the systemd mainatiners hopefully they can take a look
@flokli @Kloenk

**Comment by RaitoBezarius** [comment]
@ElvishJerricco we discussed it during one of the systemd bump, ^ — do you have an idea OTOH?

**Comment by charmoniumQ** [comment]
Link to discussion? I don't see it in the PR #265951.

**Comment by YHNdnzj** [comment]
https://github.com/NixOS/nixpkgs/pull/243242#issuecomment-1685410998

**Comment by ElvishJerricco** [comment]
@beh-10257 Well, that's a different issue than the original issue report. It sounds like the system hibernates fine, and just doesn't resume from the hibernate image.

Did you specify a `resume_offset=` in your `kernelParams`? Currently, both a `resume=` (aka the `boot.resumeDevice` nixos option) and `resume_offset=` parameter are required for hibernation to resume properly from a swap file. The scripted initrd has some fanciness for automatically finding the resume device without a `resume=` parameter, but it doesn't work for swap files. And systemd initrd needs it explicitly configured either way.

In the future, we'll be able to do much better on EFI systems with systemd-initrd. With 254, systemd can store the resume device information in EFI variables so that they're picked up by systemd-initrd automatically, without kernel parameters. When https://github.com/NixOS/nixpkgs/pull/282022 lands, I *believe* it'll work out of the box, in fact. But for now, you do need both parameters specified regardless of scripted vs systemd initrd.

**Comment by beh-10257** [comment]
@ElvishJerricco actually I though it should work out of the box since its not mentioned in the documentation which it should since I don't think these said fixes will be backbported to stable releases of nixos so maybe just add to the laptop section in the documetnation that resume and resume offset are requiered anyway gonna figure out those variables and see if it works
also should this still be open until https://github.com/NixOS/nixpkgs/pull/282022 lands

**Comment by ElvishJerricco** [comment]
Well, although the documentation probably should be improved, I dunno if that warrants an open issue claiming that it doesn't work when it does. Maybe it could be retitled?

**Comment by beh-10257** [comment]
@ElvishJerricco anyway I just tested this using `boot.resumeDevice` and `resume_offset=`
and hibernation works title is changed accordingly hopefully this bug report will become a way to see when will this be automatic I guess ??

**Comment by ChanceHarrison** [comment]
I tested automatic hibernation using #282022 (specifically by cherry picking the commit in that PR b50070eff9136082f6b3fe86129e930c62d610b6 onto nixos-unstable 9df3e30ce24fd28c7b3e2de0d986769db5d6225d) and hibernation worked for me with using systemd stage 1 (`boot.initrd.systemd.enable = true;`) without setting any resume device or offset or otherwise doing anything.

This is on a Dell XPS 15 9510, in case it becomes relevant.

(Thanks to @ElvishJerricco @ NixCon NA 2024 for the in-person support :pray:)

**Comment by SeerLite** [comment]
I tested automatic hibernation by just using
```nix
swapDevices = [
  { device = "/var/swapfile"; size = 15 * 1024; }
];
boot.initrd.systemd.enable = true;
```
on stable NixOS 24.05, since #282022 has been merged and it works too. Not sure how the configuration API should look like to make this more intuitive to use. Maybe alias `boot.initrd.systemd.enable` as something hibernation/swap-related?

**Comment by ElvishJerricco** [comment]
Yea as far as I'm aware, this can be closed. Please reopen if this remains an issue.

**Comment by imxade** [comment]
@ElvishJerricco I'm on nixos-unstable x86_64, and enabling boot.initrd.systemd is not adding resume_offset to kernel parameters

**Comment by ElvishJerricco** [comment]
@imxade It's not supposed to. When you hibernate, systemd stores the hibernation information in an EFI variable, not the kernel parameters. Then a service that runs during systemd initrd reads that info to resume from it.
